File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.py
class Solution:
    # Sorting and Hashing Solution - TC: O(nlogn), SC: O(n)
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        value_to_rank = {}  # Dictionary to store value-to-rank mapping
        sorted_list = sorted(list(set(arr)))  # Remove duplicates and sort unique elements
        
        # Assign ranks to sorted unique elements
        for index in range(len(sorted_list)): 
            value_to_rank[sorted_list[index]] = index + 1
          
        # Replace each element in the original array with its rank
        for index in range(len(arr)): 
            arr[index] = value_to_rank[arr[index]]

        return arr

File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.java
class Solution {
    // Sorting and Hashing Solution - TC: O(nlogn), SC: O(n)
    public int[] arrayRankTransform(int[] arr) {
        HashMap<Integer, Integer> numToRank = new HashMap<>();
        int[] sortedUniqueArr = Arrays.stream(arr).distinct().sorted().toArray();

        for (int i = 1; i <= sortedUniqueArr.length; i++) {
            numToRank.put(sortedUniqueArr[i-1], i);
        }

        for (int i = 0; i < arr.length; i++) {
            arr[i] = numToRank.get(arr[i]);
        }

        return arr;
    }
}